modelkit/core/model.py

```diff
@@ -473,63 +439,121 @@ def predict_batch(
+            # wrap the items with information as to their
+            # status within the cache.
+            # Whenever `cache_items` has `batch_size` elements that need
+            # to be recomputed, `_predict_cache_items` is called, which will
+            # in turn yield all results in the correct order.
+            # Once we have 2 x batch_size elements available from cache
+            # we yield them to avoid indefinite increase in the cache_item
+            # list size.
+            try:
+                if (
+                    n_items_to_compute == batch_size
+                    or n_items_from_cache == 2 * batch_size
+                ):
+                    yield from self._predict_cache_items(
+                        step, cache_items, _callback=_callback, **kwargs
+                    )
+                    cache_items = []
+                    n_items_to_compute = 0
+                    n_items_from_cache = 0
+                    step += batch_size
                 else:
-                    results.append(None)
-                    to_compute.append((kitem, key, item))
-            computed_results = self._predict_batch(
-                [item[2] for item in to_compute],
-                **kwargs,
-            )
-            for ((kitem, key, _), result) in zip(to_compute, computed_results):
-                self.redis_cache.set(key, pickle.dumps(result))
-                results[kitem] = result
-            logger.debug(
-                "Caching digest",
-                recomputed=len(computed_results),
-                from_cache=(len(results) - len(computed_results)),
-                model=self.configuration_key,
-            )
-        else:
-            # general case: items is a list of items to treat
-            # if there are multiple examples but no batching
-            # or if there are multiple examples and batching
-            results = self._validate_batch(
-                self._predict_batch(
-                    items,
-                    **kwargs,
-                ),
-                self._return_model,
-                ReturnValueValidationException,
+                    current_item = next(items)
+                    if self.cache and self.model_settings.get("cache_predictions"):
+                        if not _force_compute:
+                            # The cache will return a CacheItem with the information
+                            # as to the stored value (if it exists) and if it is missing
+                            # the cache_key needed to store it
+                            cache_item = self.cache.get(
+                                self.configuration_key, current_item, kwargs
+                            )
+                        else:
+                            # When force_recomputing, we still need to get
+                            # the cache key
+                            cache_item = CacheItem(
+                                current_item,
```

Comment:
Nit: We might name the different parameters to know what those None, True etc. refer to ?

Reply:
> Nit: We might name the different parameters to know what those None, True etc. refer to ?

We could, but it takes longer to instantiate the named tuple every time... Here the idea is to be as quick as possible. I do agree it's not so good for legibility though